### DISCUSSION TITLE
Can anyone tell me how do I decode the car file containing my user data to json in TypeScript?

### DISCUSSION THREAD

**Discussion by touhidurrr** [bluesky-social/atproto#3144] [Bluesky Client App]
This car file format is so annoying. If you want to be open, why choose an obscure file format like car with no reader available online? I tried to decode it with TS but failed. Here is my attempt with ChatGPT assistance:
```ts
const fs = require('fs');

// Decode Buffer fields in the JSON recursively
function decodeBuffers(obj) {
    if (Array.isArray(obj)) {
        return obj.map(decodeBuffers);
    } else if (obj && typeof obj === 'object') {
        if (obj.type === 'Buffer' && Array.isArray(obj.data)) {
            // Decode as a UTF-8 string or Base64 if binary
            return Buffer.from(obj.data).toString('utf-8'); 
        }
        return Object.fromEntries(
            Object.entries(obj).map(([key, value]) => [key, decodeBuffers(value)])
        );
    }
    return obj;
}

// Read input JSON
const inputFile = 'repo.json'; // Your existing JSON
const outputFile = 'human_readable_repo.json';

fs.readFile(inputFile, 'utf-8', (err, data) => {
    if (err) throw err;
    const parsed = JSON.parse(data);
    const humanReadable = decodeBuffers(parsed);

    // Write output JSON
    fs.writeFile(outputFile, JSON.stringify(humanReadable, null, 2), (err) => {
        if (err) throw err;
        console.log(`Human-readable JSON saved to ${outputFile}`);
    });
});
```
This outputs strange stuff like:
![image](https://github.com/user-attachments/assets/21d3874f-a9e1-4413-aba7-c8b2baba1d7a)

Kindly tell me how do I decode it properly.

**Comment by mary-ext** [comment]
Here's how I've dealt with it for https://boat.kelinci.net where it has the functionality to unpack CAR repositories into a tarball

https://github.com/mary-ext/boat/blob/8b0b04fdea90cf66de099fab8c0cea4478f8af45/src/views/repository/car-unpack.tsx#L52

**Comment by touhidurrr** [comment]
@mary-ext Can you write a gist or a npm package perhaps to simplify things for us, plz.
I will also suggest a name for you `car-to-tar` or `car2tar` perhaps.

**Reply by DavidBuchanan314** [reply]
If that's all you want you can just use https://boat.kelinci.net/car-unpack
